File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    N = len(begin)
    
    if target not in words:
        return 0
    
    queue = deque([]); queue.append((begin, 0))
    visit = set(); visit.add(begin)
    
    while queue:
        src, cnt = queue.popleft()
        if src == target:
            return cnt
        
        for word in words:
            check = 0
            for i in range(N):
                if word[i] == src[i]: check += 1
            if check == N-1:
                visit.add(word)
                queue.append((word, cnt+1))
        
    return 0

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <set>
#include <queue>
#include <algorithm>
using namespace std;

int N, M;

int bfs(string begin, string target, vector<string> &words){
    queue<pair<string, int>> queue;
    queue.push({begin, 0});
    
    set<string> visit;
    visit.insert(begin);
    
    while (!queue.empty()){
        string curr = queue.front().first;
        int depth = queue.front().second;
        queue.pop();
        
        if (curr == target){
            return depth;
        }
        
        for (int i=0; i<N; i++){
            int check = 0;
            for (int j=0; j<M; j++){
                if (curr[j] == words[i][j]) check++;
            }
            if (check == M-1){
                visit.insert(words[i]);
                queue.push({words[i], depth+1});
            }
        }
    }
    return 0;
}

int solution(string begin, string target, vector<string> words) {
    N = words.size();
    M = words[0].length();
    
    auto it = find(words.begin(), words.end(), target);
    if (it == words.end()) return 0;
    
    int answer = bfs(begin, target, words);
    return answer;
}